docs/refs/notations.md

```diff
@@ -1,3 +1,130 @@
 # Notation References
 
-// TODO:
+## Queries in Twoslash
+
+One of the key reasons for making Twoslash is the ability to use the TypeScript compiler to pull out type information about your code mechanically. Twoslash comes with two different ways to query your code: `?^` and `?|`.
+
+#### Extract Type
+
+Using `^?` you can pull out type information about a particular identifier in the line of code above it.
+
+```ts twoslash
+const hi = 'Hello'
+const msg = `${hi}, world`
+//    ^?
+```
+
+#### Completions
+
+Using `^|` you can pull out information about a what the auto-complete looks like at a particular location.
+
+```ts twoslash
+// @noErrors
+console.e
+//       ^|
+```
+
+What happens is that Twoslash makes a request to TypeScript to get the auto-complete at the point of the `^`, and then filters the possible outputs based on the letters following the `.`. Up to 5 results will be shown inline, and if a completion is marked as deprecated - that will be respected in the output.
+
+So, in this case, Twoslash asks TypeScript for completions of `console`, then filters down to completions which start with `e`. Note that the compiler flag for `// @noErrors` is set, because `console.e` is a failing TypeScript code sample but we don't really care about that.
+
+## Cutting a Code Sample
+
+Every Twoslash code sample needs to be a complete TypeScript program realistically, basically it needs to compile. Quite often to make it compile, there is a bunch of code which isn't relevant to the user. This can be extracted out of the code sample via `// ---cut---` which removes all of the code above it from the output.
+
+### `---cut---`
+
+Cut works after TypeScript has generated the project and pulled out all the editor information (like identifiers, queries, highlights etc) and then amends all of their offsets and lines to re-fit the smaller output. What your user sees is everything below the `---cut---`.
+
+```ts twoslash
+const level: string = 'Danger'
+// ---cut---
+console.log(level)
+```
+
+Would only show a single line.
+
+```ts twoslash
+// @filename: a.ts
+export const helloWorld: string = 'Hi'
+// ---cut---
+// @filename: b.ts
+import { helloWorld } from './a'
+
+console.log(helloWorld)
+```
+
+Would only show the last two lines, but to TypeScript it was a program with 2 files and all of the IDE information is hooked up correctly across the files. This is why `// @filename: [file]` is specifically the only Twoslash command which _is not_ removed, because if it's not relevant it can be `---cut---` away.
+
+### `---cut-after---`
```

Comment:
I also supported `---cut-start---` and `---cut-end---` pairs, which I think would be useful :) Feel free to add them, or I can do it later :)

Reply:
Yeah, open to `---cut-start---` being an "alias" to `---cut---`